Added option to specify the archive format to write + Cleanup of tabs/spaces

Saruspete opened the pull request:
By adding the archive format, this fixes a blocking bug : ability to create an archive from scratch. Otherwise, the variable "archive_format" is not set (no call to detect_archive_format).

Thus, I renamed function "detect_archive_format" to "set_archive_format" and called it when the variable is not set.

Also, I forced the refresh of "archive_format" for each "read" call, so we can merge cpio from different formats

prigaux commented:
Hi,

It seems nice :)

For "set_archive_format", i'd rather:
- keep a function "detect_archive_format" that would simply return the "format" (ODC, NewAscii, OldBinary)
- rename "set_archive_format" into "create_archive_format" that would only work with a "format"

Could you do it?

Thanks for the work!
Bonne journée :)

prigaux commented:
To be more precise, please see c7b6db906393041df79ae118bb74ce3725980243 & 6e5a051a525aec38d9cf27ed9d8264d16d8d1d4b (not tested / need testing)

Saruspete commented:
Hello,

Thanks for your update ! Your code seems fine too. 
Just a small typo and unused variable. So I added "use strict" and "use warnings" to detect this.

Maybe we should invert the hash "magics", so the type is the key, and the magic number the value. That would avoir using reverse, and still be usable for detect_archive_format (as it needs both type & magic). 

Tested with code from previous commit 0316da9

```
orig.newc.cpio:         ASCII cpio archive (SVR4 with no CRC)
orig.odc.cpio:          ASCII cpio archive (pre-SVR4 or odc)
rewrite.newc2newc.cpio: ASCII cpio archive (SVR4 with no CRC)
rewrite.newc2odc.cpio:  ASCII cpio archive (pre-SVR4 or odc)
rewrite.odc2newc.cpio:  ASCII cpio archive (SVR4 with no CRC)
rewrite.odc2odc.cpio:   ASCII cpio archive (pre-SVR4 or odc)
test.ascii.cpio:        ASCII cpio archive (SVR4 with no CRC)
test.odc.cpio:          ASCII cpio archive (pre-SVR4 or odc)
```

prigaux commented:
Reversing the hash would make it fail to handle the 2 kinds of OldBinary...

Saruspete commented:
Oh, indeed. Nevermind :)

Saruspete commented:
Do you think these commits may be merged ? Or do you need more tests ? :)

prigaux commented:
Can you redo your commits based on create_archive_format ?

Saruspete commented:
You mean the modifications done in 23cf6bb ? Or should I destroy the PR and my branch to avoid the renaming ?

prigaux commented:
You can "push -f" a new branch with clean commits. It will avoid the renaming and you will not need to create a new PR, it will replace it (afaik).

Saruspete committed:
[Fix] Normalized tabs and spaces

Removed spaces on empty lines
Removed trailing spaces
Changed tabs for 8 spaces to keep identation

Saruspete committed:
[Fix] Fixed overwrite of "mode" from optionnal hash

Saruspete committed:
[Enh] Allows to create a cpio archive by specifying the output format

By default, write will use ODC format for compability. You can set this
value by a 2nd argument to "write" sub :

$cpio = Archive::Cpio->new;
$cpio->add_data("filename.txt", "my content", { mode => 1445792815 });

$cpio->write("test.odc.cpio");
$cpio->write("test.ascii.cpio", "NewAscii");

`find test.* | cpio --quiet -H odc -o > orig.odc.cpio`;
`find test.* | cpio --quiet -H newc -o > orig.newc.cpio`;

$cpio->remove(("pouet"));
$cpio->read("orig.odc.cpio");
$cpio->write("rewrite.odc2newc.cpio", "NewAscii");
$cpio->write("rewrite.odc2odc.cpio", "ODC");

my $cpio2 = Archive::Cpio->new;
$cpio2->read("orig.newc.cpio");
$cpio2->write("rewrite.newc2newc.cpio", "NewAscii");
$cpio2->write("rewrite.newc2odc.cpio", "ODC");

Saruspete committed:
[Fix] Forced update of 'archive_format' upon reading

Saruspete commented:
Should be better now :)
